### ISSUE TITLE
Probably bug in any word hint demo

### ISSUE THREAD

**Issue by jcubic** [codemirror/CodeMirror#6653]
I was just writing hint in my code based on any [anywordhint.html](https://github.com/codemirror/CodeMirror/blob/master/demo/anywordhint.html) and it seem that the code is not correct. The regex using global flag and according to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test) test function change lastIndex of the regex test function. I'm not 100% sure if the implementation have bug, but in my code that use almost the same while loop it matched only first character of the word because lastIndex on next call was 1 and it din't much for second character. So the start while loop ended after first iteration. To fix this just remove `g` flag from the regex.

TL;DR here is the bug:
```javascript
var WORD = /[\w$]+/g, RANGE = 500;
   ...
    while (end < curLine.length && word.test(curLine.charAt(end))) ++end;
    while (start && word.test(curLine.charAt(start - 1))) --start;
```

**Comment by marijnh** [comment]
Do you have an example where it behaves incorrectly?

**Comment by jcubic** [comment]
Sorry I was not testing your code maybe it magically work but you can see the same code in my project:

https://codepen.io/jcubic/pen/abpqXvO?editors=0010

If you type `(defi` and `CTRL+Space` it much every word. The code is little bit refactored to make it cleaner but if you look at function `get_range` it have exactly same code and use regex that have g variable. Maybe your example works because it find every word in document but it's bad documentation because those while loops don't do what they suppose to do. At least this is what I thought then I tried to implement my code based on this demo.

**Comment by jcubic** [comment]
Maybe it just need a comment that explain what the while loop do, if it suppose to match only one letter.

**Comment by marijnh** [comment]
`WORD`, in this codebase, [doesn't have a `g` flag](https://github.com/codemirror/CodeMirror/blob/master/addon/hint/anyword-hint.js#L14).

**Comment by jcubic** [comment]
But demo have https://github.com/codemirror/CodeMirror/blob/master/demo/anywordhint.html#L32

**Comment by marijnh** [comment]
Ah, that's just filler content for the editor, which appears to contain an old implementation of the addon. Attached patch updates it to the current code.
